docs/specs/code/SIGNING.md

```diff
@@ -0,0 +1,424 @@
+
+For the `Headers`,
+
+- The `protected` fields MUST be empty. These are meant to be used with AEAD which we  don't need in this  specification (you can read more about it at [RFC 8152 section 5.3](https://tools.ietf.org/html/rfc8152#section-5.3)).
+- The `unprotected` field MUST contain a `enc_type` label with value `EncryptionType` so that wallets can easily know what UI to display to users.
+
+We define the encrypted payload as the union of two possibilities (defined below)
+
+```
+encrypted_message = PasswordEncryption / PubKeyEncryption
+```
+
+#### Password-based encryption
+
+For password-based encryption we don't need a receiver field (anybody who knows the password can decrypt) so we instead use the following (simplified) structure
+
+```
+COSE_Encrypt0 = [
+    Headers,
+    ciphertext : bstr / nil,
+]
+
+PasswordEncryption = 16 (COSE_Encrypt0)
+```
+
+The `Headers` MUST have `enc_type: 1` in the `unprotected` field.
+
+The `ciphertext` is simply the result of [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md) on the `signed_message`.
+
+#### Public key based encryption
+
+We only allow encrypting based on `ED25519 extended` public keys (the ones used for Cardano). To encrypt based on these public keys, you must
+
+1) Compute a password consisting of 22 case-sensitive alphanumeric (a–z, A–Z, 0–9) characters (this gives you ~128 bits of entropy)
+
+Now, for each receiver, you must
+
+2) Compute an ephemeral key pair using `ED255519 extended`
+1) Compute the shared DH secret between the private key from step (1) and the public key received (using the `exchange` functionality)
+1) Compute the `Blake2b512` hash of the secret
+1) Use this as the password to encrypt the password in (1) using [the Yoroi encryption spec](https://github.com/Emurgo/yoroi-frontend/blob/737595fec5a89409aacef827d356c9a1605515c0/docs/specs/code/ENCRYPT.md)
+
+The structure will look like the following:
+
+```
+COSE_Encrypt = [
+  Headers,
+  ciphertext : bstr / nil, # contained signed_message encrypted with random password
+  recipients : [+COSE_recipient]
+]
+
+COSE_recipient = [
+  Headers,
+  ciphertext : bstr / nil, # contains random password encrypted with shared secret
+]
+
+PubKeyEncryption = 96 (COSE_Encrypt)
+```
+
+The `Headers` for the body MUST have `enc_type: 2` in the `unprotected` field.
+
+The `Headers` for the recipient MUST have a `eph_pub: bstr` label containing the public key of the ephemeral keypair.
+
+## Multisig
+
+TODO
+
+## More headers
+
+### CWT
+
+CWT tags MAY be introduced inside `protected` headers as a CBOR map with the key `cwt`.
+
+### Payload encoding
+
+To solve `E3`, `signed_message` body header MUST `is_ascii: bool` as a `protected` header which defined whether or not the bytes the `payload` should be considered as ASCII .
```

Comment:
I only encode `is_ascii` instead of a general `encoding` header. It may be worth it to have a general `encoding` so we can do `ascii / utf8 / binary` but I figured probably supporting binary isn't really worth the trouble (could just instead sign a string of hexadecimal)